engine/src/main/java/io/camunda/zeebe/engine/processing/common/DecisionBehavior.java

```diff
@@ -0,0 +1,271 @@
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.camunda.zeebe.engine.processing.common;
+
+import static io.camunda.zeebe.util.buffer.BufferUtil.bufferAsString;
+
+import io.camunda.zeebe.dmn.DecisionEngine;
+import io.camunda.zeebe.dmn.DecisionEvaluationResult;
+import io.camunda.zeebe.dmn.EvaluatedDecision;
+import io.camunda.zeebe.dmn.EvaluatedInput;
+import io.camunda.zeebe.dmn.EvaluatedOutput;
+import io.camunda.zeebe.dmn.MatchedRule;
+import io.camunda.zeebe.dmn.ParsedDecisionRequirementsGraph;
+import io.camunda.zeebe.dmn.impl.VariablesContext;
+import io.camunda.zeebe.engine.metrics.ProcessEngineMetrics;
+import io.camunda.zeebe.engine.processing.bpmn.BpmnElementContext;
+import io.camunda.zeebe.engine.processing.streamprocessor.writers.StateWriter;
+import io.camunda.zeebe.engine.state.deployment.PersistedDecision;
+import io.camunda.zeebe.engine.state.deployment.PersistedDecisionRequirements;
+import io.camunda.zeebe.engine.state.immutable.DecisionState;
+import io.camunda.zeebe.engine.state.immutable.VariableState;
+import io.camunda.zeebe.engine.state.immutable.ZeebeState;
+import io.camunda.zeebe.protocol.impl.encoding.MsgPackConverter;
+import io.camunda.zeebe.protocol.impl.record.value.decision.DecisionEvaluationRecord;
+import io.camunda.zeebe.protocol.impl.record.value.decision.EvaluatedDecisionRecord;
+import io.camunda.zeebe.protocol.impl.record.value.decision.MatchedRuleRecord;
+import io.camunda.zeebe.protocol.record.intent.DecisionEvaluationIntent;
+import io.camunda.zeebe.protocol.record.value.ErrorType;
+import io.camunda.zeebe.stream.api.state.KeyGenerator;
+import io.camunda.zeebe.util.Either;
+import io.camunda.zeebe.util.buffer.BufferUtil;
+import java.io.ByteArrayInputStream;
+import java.util.stream.Collectors;
+import org.agrona.DirectBuffer;
+
+public class DecisionBehavior {
+
+  private static final DecisionInfo UNKNOWN_DECISION_INFO = new DecisionInfo(-1L, -1);
+  private final DecisionEngine decisionEngine;
+  private final DecisionState decisionState;
+  private final VariableState variableState;
+  private final ProcessEngineMetrics metrics;
+  private final StateWriter stateWriter;
+  private final KeyGenerator keyGenerator;
+
+  public DecisionBehavior(
+      final DecisionEngine decisionEngine,
+      final ZeebeState zeebeState,
+      final ProcessEngineMetrics metrics,
+      final StateWriter stateWriter,
+      final KeyGenerator keyGenerator) {
+    this.decisionEngine = decisionEngine;
+    decisionState = zeebeState.getDecisionState();
+    variableState = zeebeState.getVariableState();
+    this.metrics = metrics;
+    this.stateWriter = stateWriter;
+    this.keyGenerator = keyGenerator;
+  }
+
+  public Either<Failure, DecisionEvaluationResult> evaluateDecision(
+      final String decisionId, final long scopeKey, final BpmnElementContext bpmnContext) {
```

Comment:
🔧 I'd expect the variables will need to passed to this method directly. Currently we fetch them from the state, which won't work with the standalone evaluation.

Reply:
In that case, I think I'll need to decouple the logic further.

### Problem

* The `scopeKey` is used only for providing better `Failure` data. These types of failures are created when:
  * There are issues with finding or parsing the Decision or DRG. 
  * The evaluation failed.
  * When we create a Decision parsing `Failure`, we also use `ErrorType.CALLED_DECISION_ERROR`. This wouldn't be correct anymore. So we would need to make this part more flexible. I'm also wondering what would be the `ErrorType` when a Decision or DRG can't be found/parsed during a standalone Decision evaluation. But I don't have to tackle this now.
* We use the `bpmnContext` to add BPMN process data to a `DecisionEvaluationEvent`.
  * In the case of a standalone Decision evaluation, these fields would be empty.

### Requirements
So there are three points where the `BpmnDecisionBehavior` will differ from the `DecisionBehavior`:
1. There is an issue with finding the Decision or DRG. The `Failure` should be different depending on the caller (Business Rule task or standalone).
2. The evaluation fails. The `Failure` should include a `scopeKey` only when a Business Rule task is used.
3. After the Decision evaluation, a `DecisionEvaluationEvent` is written. When a Business Rule task is used, the appropriate BPMN process data needs to be included.

### Possible solutions
For both points, we would need to let the `BpmnDecisionBehavior` adjust some data, or do some extra steps. I can think of the following ways to do this:
1. Allow the `DecisionBehavior` to be passed three `function`s. 
    1. The first would define how a `Failure` is defined when there is a Decision/DRG finding/parsing failure.
    2. The second would define how a `Failure` is defined when there is a Decision evaluation failure.
    3. The third would define additional adjustments to an existing `DecisionEvaluationEvent` before writing it to the state.
2. Decouple the logic inside the `evaluateDecision(...)` method into smaller `public` methods, that can be used by the `BpmnDecisionBehavior`. The `public` methods should map (at least) to the following three states:
    1. Find a Decision by ID and locate the DRG.
    2. Evaluate the Decision
    3. Write a `DecisionEvaluationEvent` and metrics.

### Decision

I would go with the second solution, i.e. decouple the logic into smaller `public` methods, for the following reasons.

1. It provides a cleaner separation of "standalone" DMN concepts vs additional BPMN ones.
2. It allows it to be extended arbitrarily instead of providing specific fields to modify its behavior.

Reply:
@koevskinikola I agree with your conclusion. I also prefer option 2. 

Regarding the `DecisionEvaluationContext`. I wouldn't put the `evaluationResult` in there. If we do we have this context object where we can get a result, and this will be null at the beginning. At some arbitrary point in the code this will be set to a value. Using the context requires more in-depth knowledge about the methods that take this context a parameter. I don't like that.
That leaves just the `Decision` and the `DRG` that would be part of it. I'd argue it's overkill to have a specific class to combine 2 parameters into one. But I'll leave that decision to you. I wouldn't reject a PR because of it 😄